packages/server/src/utils/LocalFileSystemHandler.js

```diff
@@ -0,0 +1,124 @@
+    }
+
+    createDirectoryIfNotExists(dir) {
+        if (!fs.existsSync(dir)) {
+            fs.mkdirSync(dir, { recursive: true, mode: '0744' });
+        }
+    }
+
+
+    saveFile(directoryPath, filename, buffer) {
+        return new Promise((resolve, reject) => {
+            try {
+                this.verifyPath(directoryPath)
+                this.createDirectoryIfNotExists(path.join(this._uploadDirectory, directoryPath))
+            } catch ( err ) {
+                reject(err)
+            }
+            
+            fs.writeFile(path.join(this._uploadDirectory, directoryPath, filename), buffer, (err) => {
+                if ( err ) {
+                    reject(err)
+                } else {
+                    resolve()
+                }
+            })
+        })
+    }
+
+    /**
+     * Downloads the binary data for a file
+     * @param {*} directoryPath
+     * @param {*} filename
+     */
+    getFile(directoryPath, filename) {
+        return new Promise((resolve, reject) => {
+            try {
+                this.verifyPath(directoryPath)
+            } catch ( err ) {
+                reject(err)
+            }
+
+            fs.readFile(path.join(this._uploadDirectory, directoryPath, filename), (err, data) => {
+                if ( err ) {
+                    console.log(err)
+                    reject(err)
+                } else {
+                    resolve({ Body:data })
+                }
+            })
+        })
+    }
+
+    /**
+     * Deletes a single file with the specified path and filename
+     * @param {*} directoryPath
+     * @param {*} filename
+     */
+    deleteFile(directoryPath, filename) {
+        return new Promise((resolve, reject) => {
+            try {
+                this.verifyPath(directoryPath)
+            } catch ( err ) {
+                reject(err)
+            }
+
+            fs.unlink(path.join(directoryPath, filename), (err) => {
+                if ( err ) {
+                    reject(err)
+                } else {
+                    resolve()
+                }
+            })
+        })
+    }
+
+    deleteDir(directoryPath) {
+        return new Promise((resolve, reject) => {
+            try {
+                this.verifyPath(directoryPath)
+            } catch ( err ) {
+                reject(err)
+            }
+
+            rimraf(path.join(this._uploadDirectory, directoryPath), (err) => {
+                if ( err ) { 
+                    reject(err)
+                } else {
+                    resolve()
+                }
+            })
+        })
+    }
+
+    verifyPath(directoryPath) {
+        if ( directoryPath.indexOf('..') !== -1 ) {
+            throw new Error('Relative paths not allowed')
+        }
+
+        if ( directoryPath.indexOf('~') === 0 ) {
+            throw new Error('Tilde not allowed in directoryPath')   //avoid reading from current user directory
+        }
+
+        if ( directoryPath.indexOf('/') === 0 ) {
```

Comment:
I'm confused, what kind of path can it be if it can't be relative and can't be absolute?

Reply:
it should be relative but not able to escape up to it's parent outside of the allowed directory using '..', that's what I meant with the description at least, or if someone somehow manages to pass down an absolute path it should not be able to save/read files to other places on disk